server/src/main/java/org/opensearch/action/admin/indices/shrink/TransportResizeAction.java

```diff
@@ -250,6 +258,39 @@ static CreateIndexClusterStateUpdateRequest prepareCreateIndexRequest(
             .copySettings(resizeRequest.getCopySettings() == null ? false : resizeRequest.getCopySettings());
     }
 
+    /**
+     * Calculate target index's shards number according to maxShardSize and the source index's storage(only primary shards included)
+     * @param maxShardSize the maximum size of a primary shard in the target index
+     * @param sourceIndexShardStoreStats primary shards' store stats of the source index
+     * @param sourceIndexMetaData source index's metadata
+     * @return target index's shards number
+     */
+    protected static int calculateTargetIndexShardsNum(
+        ByteSizeValue maxShardSize,
+        StoreStats sourceIndexShardStoreStats,
+        IndexMetadata sourceIndexMetaData
+    ) {
+        if (maxShardSize == null
+            || sourceIndexShardStoreStats == null
+            || maxShardSize.getBytes() == 0
+            || sourceIndexShardStoreStats.getSizeInBytes() == 0) {
+            return 1;
+        }
+
+        int sourceIndexShardsNum = sourceIndexMetaData.getNumberOfShards();
+        int minValue = (int) Math.ceil((double) sourceIndexShardStoreStats.getSizeInBytes() / maxShardSize.getBytes());
+        if (minValue >= sourceIndexShardsNum) {
```

Comment:
It took me some time to understand this condition (sorry for that), please correct me if I am wrong, but: if the resize request in shrink mode needs more shards (than source index), we keep the amount of shards unchanged, is that right? Could we the comment to explain the algorithm for future references? Thank you.